### ISSUE TITLE
Additional reformat option for spaces between headings

### ISSUE THREAD

**Issue by danhaywood** [asciidoctor/asciidoctor-intellij-plugin#428]
I love Code > Reformat Code to clean up my asciidoc pages.  However, after reformatting there is only a single line before all paragraphs and headings.

For myself, I prefer a little bit more "air" before my headings, two or three lines, say.

Perhaps this could be made into a setting that can be tweaked.

**Comment by ahus1** [comment]
I'm labeling this as a "good first issue". I'm happy to help along the way. 

Starting point to add a settings for this would be AsciiDocLanguageCodeStyleSettingsProvider. Then the user can configure this via the settings in "Code Style > AsciiDoc" or via an .editorconfig file. 

Once the configuration option is available, the formatting in AsciiDocBlock can pick it up like it already picks up the configuration for ONE_SENTENCE_PER_LINE.

There are some existing tests regarding formatting in AsciiDocFormattingModelBuilderTest. The test files are located in folder testData/actions/reformatFileInEditor.

**Comment by danhaywood** [comment]
It'd be fun to do some stuff on this, and I've built tools in the past
using Eclipse ... but at the moment I'm overloaded with trying to make some
progress on my own open source project :-(

If it's any consolation, I do appreciate how difficult it is to get people
to contribute rather than merely consume!

Maybe/hopefully in a few months, would be nice to take a break and explore
into this...

**Comment by ahus1** [comment]
@danhaywood - no worries; still I need to make sure that this project is not a one-person-show. Having more people contributing would allow us to build a better product, and it would be way more fun. Therefore I'm marking some tickets "good first issue".

**Comment by mminot-yseop** [comment]
I agree with this. Generally speaking, I always disable formatting in this extension (and recommend to my colleagues to do the same, to avoid mayhem and needlessly long diffs in pull requests), as I think the default behavior tends to lead to cluttered-looking AsciiDoc files.

Regarding headings, the rule we try to apply in my company is:

```asciidoc
= Title

Preamble.


== Foo

=== Bar

Plop


=== Plup

Palala
```

In other words: **two blank lines instead of one** whenever a heading is **preceded by anything else than a heading**. This way the source code makes the structure a lot clearer IMO. The extension squeezes all that into some kind of zebra-crossing-looking list where my brain struggles to find out what is below what.

Similarly, I’m always surprised to see that `[source, bash]` gets turned to a space-less `[source,bash]`. It’s really confusing when you’re used to working with programing languages where having a space after commas when allowed is basically the norm. 🤔

**Comment by mminot-yseop** [comment]
> I don't agree at all with using two blank lines before headings. […] That just makes the AsciiDoc source inconsistent.

I kinda get that point of view as well, but I was surprised to see the extension had a setting for blank lines _after_ headings and not _before_ xD I never saw anyone wanting boatloads of spaces after those. And as the maintainer noted, the implementation could be similar for both settings, and then lots of people could tune it the way they want (organization-wide, like we already often do for Java, etc.) and be happy.

But basically, automatic formatting (all the more when it’s turned on by default) without sufficient customization options often turns out to be more of an annoyance than anything, unless it’s for something like JSON, etc. where people don’t care much about anything as long as the formatting is consistent. 🤷🏻‍♀️

> Study AsciiDoc source files almost anywhere

I take things like that with a grain of salt, though. Typically, 99% of the Bash I see on the web is unreadable and non-robust, and people tend to write `sed` and `awk` scripts without newlines nor spaces for absolutely no reason. Widespread stuff often stems from laziness and sloppiness more than from logic and good intentions.

**Comment by mminot-yseop** [comment]
> > I’m always surprised to see that `[source, bash]` gets turned to a space-less `[source,bash]`.
> 
> This one I think could be a preference. I've done it both ways and I still can't settle on which one I prefer. My instinct is to prefer less characters, but I understand other other people have the opposite instinct.

I also find it weird / funny how inconsistent the behavior towards spaces currently is. For example, if you write…

```asciidoc
[ cols   =  "<.^m,  d,  m", options  =  header ]
```

… the formatter will obliterate the post-comma space but won’t care one bit about the others. 🤔 So at the end of the day, if you’re the kind of developer to put one space on each side of `=`, you end up with more spaces _within_ each key-value pair than _between_ them. It feels off-putting to me.

**Comment by mojavelinux** [comment]
> Widespread stuff often stems from laziness and sloppiness more than from logic and good intentions.

Well, I work extremely hard to advocate for best practices for AsciiDoc and a lot of the AsciiDoc out there is the way it is because I have communicated these conventions. So it's not just random sloppiness. It's very intentional.

> the formatter will obliterate the post-comma space but won’t care one bit about the others.

It seems reasonable to me that the formatter would remove insignificant spaces. Again, whether the space is kept after the comma is something I think could be configurable.

**Comment by mojavelinux** [comment]
I feel like you're being rather accusatory rather than trying to politely communicating the behavior you think the formatter should have. Okay, so the formatter may not be catching all the spaces. It's not perfect. If it was, there wouldn't be an issue tracker. You can say "I notice that the formatter is not removing all insignificant spaces in the attribute list. Would it be possible to implement it that way?" Saying "this is off-putting" is disrespectful and that tone has no place in this community.
